csi.proto

```diff
@@ -502,6 +556,213 @@ message ControllerServiceCapability {
     RPC rpc = 1;
   }
 }
+message CreateSnapshotRequest {
+  // The ID of the source volume to be snapshotted.
+  // This field is REQUIRED.
+  string source_volume_id = 1;
+
+  // The suggested name for the snapshot. This field is REQUIRED.
+  // It serves two purposes:
+  // 1) Idempotency - This name is generated by the CO to achieve
+  //    idempotency. If `CreateSnapshot` fails, the snapshot may or
+  //    may not be created. In this case, the CO may call
+  //    `CreateSnapshot` again, with the same name, to ensure the
+  //    snapshot exists. The Plugin should ensure that multiple
+  //    `CreateSnapshot` calls for the same name do not result in more
+  //    than one snapshot created corresponding to that name. If a
+  //    Plugin is unable to enforce idempotency, the CO's error recovery
+  //    logic could result in multiple (unused) snapshots being created.
+  // 2) Suggested name - Some storage systems allow callers to specify
+  //    an identifier by which to refer to the newly created snapshot.
+  //    If a storage system supports this, it can optionally use this
+  string name = 2;
+
+  // The description of the snapshot.
+  // This field is OPTIONAL.
+  string description = 3;
+
+  // Secrets required by plugin to complete snapshot creation request.
+  // A secret is a string to string map where the key identifies the
+  // name of the secret (e.g. "username" or "password"), and the value
+  // contains the secret data (e.g. "bob" or "abc123").
+  // Each key MUST consist of alphanumeric characters, '-', '_' or '.'.
+  // Each value MUST contain a valid string. An SP MAY choose to accept
+  // binary (non-string) data by using a binary-to-text encoding scheme,
+  // like base64.
+  // An SP SHALL advertise the requirements for required secret keys and
+  // values in documentation.
+  // CO SHALL permit passing through the required secrets.
+  // A CO MAY pass the same secrets to all RPCs, therefore the keys for
+  // all unique secrets that an SP expects must be unique across all CSI
+  // operations.
+  // This information is sensitive and MUST be treated as such (not
+  // logged, etc.) by the CO.
+  // This field is OPTIONAL.
+  map<string, string> controller_create_secrets = 4;
+
+  // Plugin specific parameters passed in as opaque key-value pairs.
+  // This field is OPTIONAL. The Plugin is responsible for parsing and
+  // validating these parameters. COs will treat these as opaque.
+  // Use cases for opaque parameters:
+  // - Specify a policy to automatically clean up the snapshot.
+  // - Specify an expiration date for the snapshot.
+  // - Specify whether the snapshot is readonly or read/write.
+  // - Specify if the snapshot should be replicated to some place.
+  // - Specify primary or secondary for replication systems that
+  //   support snapshotting only on primary.
+  map<string, string> parameters = 5;
+
+  // NOTE: In the future, we may want to add “tags” to identify a
```

Comment:
I don't think we need the note in here? Sounds like this should be an issue on the repo.